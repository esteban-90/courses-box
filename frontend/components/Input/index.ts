export * from './Input'
export * from './Feedback'
