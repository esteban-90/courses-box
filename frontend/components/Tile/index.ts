export * from './Tile'
export * from './CenteredTile'
