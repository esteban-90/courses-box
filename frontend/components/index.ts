export * from './Button'
export * from './Checkbox'
export * from './Course'
export * from './Icon'
export * from './IconButton'
export * from './Input'
export * from './Layout'
export * from './Link'
export * from './Logo'
export * from './Switch'
export * from './Tile'
