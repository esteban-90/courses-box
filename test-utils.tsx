import { FC, ReactElement, ReactNode } from 'react'
import { render, RenderOptions, RenderResult } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'
import { Themes } from '@/styles/themes'

const Wrapper: FC<{ children: ReactNode }> = ({ children }): JSX.Element => {
  return <ThemeProvider theme={Themes.light}>{children}</ThemeProvider>
}

const customRender = (ui: ReactElement, options?: RenderOptions): RenderResult => {
  return render(ui, { wrapper: Wrapper, ...options })
}

export * from '@testing-library/react'
export { customRender as render }

// 14
