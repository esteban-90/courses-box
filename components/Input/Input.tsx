import { FC, ChangeEventHandler, InputHTMLAttributes, ReactNode } from 'react'
import styled from '@emotion/styled'
import { Icon, AvailableIcons } from '@/components/Icon'
import { makeShadow } from '@/styles/helpers'

const StyledInput = styled.input`
  all: unset;
  border-radius: 1rem;
  font-size: 1.4rem;
  padding: 0 1.4rem;

  width: 100%;
  height: 100%;
  color: ${({ theme }) => theme.font.regular};

  box-shadow: ${({ theme }) => {
    const { shadow1, shadow2 } = theme.components
    const $shadow = makeShadow(shadow1, shadow2, true)
    return $shadow
  }};

  &:focus {
    box-shadow: ${({ theme }) => {
      const { shadow1, shadow2 } = theme.components
      const $shadow = makeShadow(shadow1, shadow2)
      return $shadow
    }};
  }

  &::placeholder {
    opacity: 0.7;
    color: ${({ theme }) => theme.font.placeholder};
  }

  &:not(:placeholder-shown) {
    & ~ svg {
      display: none;
    }
  }

  &:hover {
    cursor: text;
  }
`

const StyledWrapper = styled.div`
  display: flex;
  align-items: center;
  position: relative;
  width: 100%;
  height: 100%;
`

type StyledLabelProps = {
  /** Input height */
  height?: number

  /** Input width */
  width?: number
}

const StyledLabel = styled.label<StyledLabelProps>`
  display: flex;
  justify-content: flex-start;
  flex-direction: column;
  font-size: 1rem;
  height: ${({ height }) => height}rem;
  width: ${({ width }) => width}rem;
  color: ${({ theme }) => theme.font.regular};
`

const StyledIcon = styled(Icon)`
  position: absolute;
  right: 1.4rem;
  opacity: 0.7;
  color: ${({ theme }) => theme.font.placeholder};
`

const StyledText = styled.span`
  padding-left: 1.4rem;
`

type InputProps = {
  /** Input placeholder */
  placeholder: string

  /** Change handler */
  onChange: ChangeEventHandler<HTMLInputElement>

  /** Input label */
  label?: string

  /** Icon */
  icon?: AvailableIcons

  /** Feedback for input */
  feedback?: ReactNode
} & StyledLabelProps &
  InputHTMLAttributes<HTMLInputElement>

export const Input: FC<InputProps> = ({
  label,
  icon,
  feedback,
  height = 7,
  width = 20,
  className,
  ...rest
}): JSX.Element => {
  return (
    <StyledLabel height={height} width={width} className={className}>
      {label && <StyledText>{label}</StyledText>}

      <StyledWrapper>
        <StyledInput {...rest} />
        {icon && <StyledIcon name={icon} />}
      </StyledWrapper>

      {feedback && <StyledText>{feedback}</StyledText>}
    </StyledLabel>
  )
}
