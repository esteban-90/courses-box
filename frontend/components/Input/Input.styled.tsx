import styled from '@emotion/styled'
import { Icon as _Icon } from '@/components/Icon'
import { makeShadow } from '@/styles/helpers'

export const Label = styled.span`
  grid-area: label;
  padding-left: 1.4rem;
`

export const Feedback = styled(Label)`
  grid-area: feedback;
`

export const Icon = styled(_Icon)`
  position: absolute;
  right: 1.4rem;
  opacity: 0.7;
  color: ${({ theme }) => theme.font.placeholder};
`

export const InnerWrapper = styled.div`
  grid-area: input;
  display: flex;
  align-items: center;
  position: relative;
  width: 100%;
  height: 100%;
`

export const Input = styled.input`
  all: unset;
  border-radius: 1rem;
  font-size: 1.4rem;
  padding: 0 1.4rem;

  width: 100%;
  height: 100%;
  color: ${({ theme }) => theme.font.regular};

  box-shadow: ${({ theme }) => {
    const { shadow1, shadow2 } = theme.components
    const boxShadow = makeShadow(shadow1, shadow2, true)
    return boxShadow
  }};

  &:focus {
    box-shadow: ${({ theme }) => {
      const { shadow1, shadow2 } = theme.components
      const boxShadow = makeShadow(shadow1, shadow2)
      return boxShadow
    }};
  }

  &::placeholder {
    opacity: 0.3;
    color: ${({ theme }) => theme.font.placeholder};
  }

  &:not(:placeholder-shown) {
    & ~ svg {
      display: none;
    }
  }

  &:hover {
    cursor: text;
  }
`

export type WrapperProps = {
  /** Input height */
  height?: number
  /** Input width */
  width?: number
  /** Label visibility */
  isLabelVisible?: boolean
  /** Feedback visibility */
  isFeedbackVisible?: boolean
}

export const OuterWrapper = styled.label<WrapperProps>`
  font-size: 1rem;
  display: grid;
  gap: 0.1rem;
  grid-template-areas:
    'label'
    'input'
    'feedback';

  grid-template-rows: ${({ isLabelVisible, isFeedbackVisible }) => {
    if (isLabelVisible && isFeedbackVisible) return '1fr 3fr 1fr'
    if (isLabelVisible) return '1fr 4fr 0fr'
    if (isFeedbackVisible) return '0fr 4fr 1fr'
    return '0fr 1fr 0fr'
  }};

  height: ${({ height }) => height}rem;
  width: ${({ width }) => width}rem;
  color: ${({ theme }) => theme.font.regular};
`
