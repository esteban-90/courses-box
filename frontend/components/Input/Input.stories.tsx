import type { ComponentStoryObj, ComponentMeta } from '@storybook/react'
import { screen, userEvent } from '@storybook/testing-library'
import { Input, Feedback } from '@/components/Input'

export default {
  title: 'Controls/Input',
  component: Input,
  args: {},
} as ComponentMeta<typeof Input>

export const PrimaryInput: ComponentStoryObj<typeof Input> = {
  async play() {
    await userEvent.type(screen.getByRole('textbox'), 'String')
  },

  args: {
    placeholder: 'Your name',
    label: 'Name:',
  },
}

export const WithIcon: ComponentStoryObj<typeof Input> = {
  args: {
    icon: 'Search',
    placeholder: 'Search',
    height: 4,
  },
}

export const WithValidFeedback: ComponentStoryObj<typeof Input> = {
  args: {
    placeholder: 'Some text',
    label: 'Text:',
    feedback: <Feedback isValid>Looks good!</Feedback>,
  },

  argTypes: {
    feedback: {
      control: false,
    },
  },
}

export const WithInvalidFeedback: ComponentStoryObj<typeof Input> = {
  args: {
    placeholder: 'Some text',
    feedback: <Feedback>Required!</Feedback>,
  },

  argTypes: {
    feedback: {
      control: false,
    },
  },
}
