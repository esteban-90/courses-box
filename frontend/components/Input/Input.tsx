import type { FC, ChangeEventHandler, InputHTMLAttributes, ReactNode, ForwardedRef } from 'react'
import { forwardRef } from 'react'
import type { IconNames } from '@/components/Icon'
import * as Styled from '@styled/Input'

type InputProps = {
  /** Input placeholder */
  placeholder: string
  /** Change handler */
  onChange?: ChangeEventHandler<HTMLInputElement>
  /** Input label */
  label?: string
  /** Icon */
  icon?: IconNames
  /** Feedback for input */
  feedback?: ReactNode
} & Styled.WrapperProps &
  InputHTMLAttributes<HTMLInputElement>

export const Input: FC<InputProps> = forwardRef(
  ({ label, icon, feedback, height = 7, width = 20, className, ...rest }, ref): JSX.Element => {
    return (
      <Styled.OuterWrapper
        height={height}
        width={width}
        isLabelVisible={!!label}
        isFeedbackVisible={!!feedback}
        className={className}
      >
        <Styled.Label>{label}</Styled.Label>
        <Styled.InnerWrapper>
          <Styled.Input ref={ref as ForwardedRef<HTMLInputElement>} {...rest} />
          {icon && <Styled.Icon name={icon} />}
        </Styled.InnerWrapper>
        <Styled.Feedback role='alert'>{feedback}</Styled.Feedback>
      </Styled.OuterWrapper>
    )
  }
)

Input.displayName = 'Input'
