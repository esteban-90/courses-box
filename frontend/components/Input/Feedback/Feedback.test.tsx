import { render } from '@/test-utils'
import { Feedback } from '@/components/Input'

describe('Feedback test cases', () => {
  it('Render check for valid', () => {
    const element = <Feedback isValid>Looks good!</Feedback>
    const { asFragment } = render(element)

    expect(asFragment()).toMatchSnapshot()
  })

  it('Render check for invalid', () => {
    const element = <Feedback>Please provide a valid value</Feedback>
    const { asFragment } = render(element)

    expect(asFragment()).toMatchSnapshot()
  })
})
