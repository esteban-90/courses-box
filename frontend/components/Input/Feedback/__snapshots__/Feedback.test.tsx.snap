// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback test cases: should check render for invalid feedback 1`] = `
<DocumentFragment>
  <span
    class="css-xjluvf"
  >
    Please provide a valid value
  </span>
</DocumentFragment>
`;

exports[`Feedback test cases: should check render for valid feedback 1`] = `
<DocumentFragment>
  <span
    class="css-116yo8b"
  >
    Looks good!
  </span>
</DocumentFragment>
`;
